opencog/cython/PythonEval.cc

```diff
@@ -528,10 +544,182 @@ Handle PythonEval::apply(const std::string& func, Handle varargs)
+                (char*) "truth_value_ptr_object", NULL);
+
+        // Make sure we got a truth value pointer.
+        pyError = PyErr_Occurred();
+        if (pyError) {
+            throw RuntimeException(TRACE_INFO, 
+                "Python function '%s' did not return TruthValue!",
+                funcName.c_str());
+        }
+
+        // Get the pointer to the truth value pointer. Yes, it does
+        // contain a pointer to the shared_ptr not the underlying.
+        tvpPtr = static_cast<TruthValuePtr*>
+                (PyLong_AsVoidPtr(pyTruthValuePtrPtr));
+
+        // Assign the truth value pointer using this pointer before
+        // we decrement the reference to pyTruthValue since that
+        // will delete this pointer.
+        tvp = *tvpPtr;
+
+        logger().info() << "TruthValue from : " + moduleName << " is " <<
+                tvp->toString();
+
+        // Cleanup the reference count.
+        Py_DECREF(pyTruthValuePtrPtr);
+    } 
+
+    if (pyError) {
+        // Error - save the message.
+        errorCallingFunction = true;
+
+        // Construct the error message.
+        PyObject *pyErrorType, *pyErrorMessage, *pyTraceback;
+        PyErr_Fetch(&pyErrorType, &pyErrorMessage, &pyTraceback);
+        std::stringstream errorStringStream;
+        errorStringStream << "Python error in " << func.c_str();
+        if (pyErrorMessage) {
+            char* errorMessage = PyString_AsString(pyErrorMessage);
+            if (errorMessage) {
+                errorStringStream << ": " << errorMessage << ".";
+            }
+            
+            // NOTE: The traceback can be NULL even when the others aren't.
+            Py_DECREF(pyErrorType);
+            Py_DECREF(pyErrorMessage);
+            if (pyTraceback)
+                Py_DECREF(pyTraceback);
+        }
+        errorString = errorStringStream.str();
+
+        // PyErr_Occurred returns a borrowed reference, so don't do this:
+        // Py_DECREF(pyError);
+    }
+    
+    // Cleanup the reference counts for Python objects we no longer reference.
+    // Since we promoted the borrowed pyExecuteApplyTVFunc reference, we need to
+    // decrement it here.
+    Py_DECREF(pyExecuteApplyTVFunc);
+    Py_DECREF(pyArgs);
+
+    // In case PyObject_CallObject returns an error, we need to check for NULL.
+    if (pyTruthValue)
+        Py_DECREF(pyTruthValue);
+
+    // NOTE: PyTuple_SetItem "steals" the reference to the object that
+    // is set for the list item. That means that the list assumes it
+    // owns the references after the call. So don't do this:
+    // Py_DECREF(pyUserFunc);
+    // Py_DECREF(pyUUID);
+
+    // Release the GIL. No Python API allowed beyond this point.
+    PyGILState_Release(gstate);
+
+    // Throw on error.
+    if (errorCallingFunction)
+        throw RuntimeException(TRACE_INFO, errorString.c_str());
+
+    // Return the truth value pointer from the function call.
+    return tvp;
```

Comment:
Gahh!!! A giant cut-n-pate of the code immediately above, with assorted tweaks to handle TV instead of handle.  Surely it would have been better to abstract the common parts away, and just have a little kernel that holds only the differences? I think you must be getting tired of this game. ... but I think you are just about out of that long tunnel, anyway...